### PR TITLE
Puyolib: Basic console debug

### PR DESCRIPTION
**Pull request by LossRain**
This request adds basic console output for all code with access to the ppvs::Game object in Puyolib, allowing easier debugging of internal Puyolib structures without a debugger.

This commit is client-agnostic as it allows any client to implement its own debug handling and uses std::string as a transport vehicle. Yet, this change might be breaking for ClientNG as I wasn't able to compile it successfully to test this scenario. 

There are currently no flags or methods to explicitly disable this information display outside of Qt's own switches.

### CODE REVIEW

File: Client/gamemanager.cpp [hunk 1]
```diff
@@ -9,10 +9,34 @@
...
+#include <QObject>
 
 volatile bool loopEnabled;
 
+/* Handles the Puyolib debug information transfer. See Puyolib/DebugLog.h for more info. */
+void GameManager::handleDebugLog(const std::basic_string<char>& text, ppvs::DebugMessageType severity)
+{
+	switch (severity) {
+	case ppvs::DebugMessageType::DEBUG:
+		qDebug().noquote() << "Puyolib:" << QString(text.c_str()) << '\n';
```
Comment (jchv) [inline Client/gamemanager.cpp:23]: It'd probably be easier to use the printf-style variants here, something like:

```cpp
    qDebug("Puyolib: %s\n", text);
```

File: Client/gamemanager.cpp [hunk 2]
```diff
@@ -9,10 +9,34 @@
...
 #include "settings.h"
 #include <QApplication>
+#include <QDebug>
 #include <QMessageBox>
+#include <QObject>
 
 volatile bool loopEnabled;
 
+/* Handles the Puyolib debug information transfer. See Puyolib/DebugLog.h for more info. */
+void GameManager::handleDebugLog(const std::basic_string<char>& text, ppvs::DebugMessageType severity)
```
Comment (jchv) [inline Client/gamemanager.cpp:19]: This should probably be
`const std::basic_string<char>&` -> `const std::string&`

File: Client/gamemanager.cpp [hunk 3]
```diff
@@ -109,6 +134,13 @@ ppvs::RuleSetInfo GameManager::createRules()
 
 	return rs;
 }
+/* Creates a new instance of DebugLog */
+ppvs::DebugLog* GameManager::createDebug()
+{
+	dbg = new ppvs::DebugLog;
+	dbg->setLogHandler([this](std::basic_string<char> text, ppvs::DebugMessageType sev) { this->handleDebugLog(text, sev); });
```
Comment (jchv) [inline Client/gamemanager.cpp:141]: Like above, this should probably be
`std::basic_string<char>` -> `std::string`

File: Puyolib/DebugLog.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ * Before calling: make sure that the logging handler is deployed and bound.
+ * */
+void DebugLog::log(std::string text, DebugMessageType severity = DebugMessageType::DEBUG)
+{
+	if (m_log_handler != NULL) {
+		m_log_handler(text, severity);
+	}
+}
+
+} // ppvs
```
Comment (jchv) [inline Puyolib/DebugLog.cpp:41]: Nit: missing newline before EOF here, best to add it so Git is happy.

File: Puyolib/Game.cpp [hunk 5]
```diff
@@ -56,6 +59,8 @@ Game::~Game()
 	delete m_statusFont;
 	delete m_data->front;
 	delete m_data;
+	delete m_debug;
+	m_debug = nullptr;
```
Comment (jchv) [inline Puyolib/Game.cpp:63]: I can see why you would do this, but it is probably a bit unnecessary, since we don't clear any of the other pointers after deleting them in the dtor. There's not a whole lot involving this object that can happen after this that would be valid/defined behavior anyway.

File: Puyolib/DebugLog.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	INFO = 4,
+	DEBUG = 8
+};
+
+/* A more precise debug message class, might be used in exception handling and other scenarios
+ * where we don't want to handle raw strings.
+ * */
+class DebugMessage {
+public:
+	DebugMessage(std::basic_string<char> text, DebugMessageType severity = DebugMessageType::NONE) : m_text(text), m_severity(severity) {};
```
Comment (jchv) [inline Puyolib/DebugLog.h:26]: Another instance of `std::basic_string<char>` that should most likely just be `std::string`

File: Puyolib/DebugLog.h [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+ * */
+class DebugLog {
+public:
+	DebugLog() = default;
+	~DebugLog();
+
+	void log(DebugMessage *msg);
+	void log(std::string text, DebugMessageType severity);
+
+	void setLogHandler(std::function<void(std::basic_string<char>,DebugMessageType)> f);
```
Comment (jchv) [inline Puyolib/DebugLog.h:48]: Another instance of `std::basic_string<char>` that should probably be `std::string`. Also, it might be a good idea to set a `using` alias for this `std::function` type, since inline `std::function` types are a bit wordy and kind of suck to repeat.

**Review by jchv — CHANGES_REQUESTED** [review]
Overall not too bad. I think the design is alright and we can generally move ahead to get it merged.

The provision for future structured logging with `DebugMessage` feels a little premature (and also, I feel like we shouldn't try to overcomplicate text logging too much, as we can always develop _other_ debug tools later) and I feel we should probably implement standard C++ tools like the newer `fmt` library (or the version that was added to C++20, `std::format`) but I am OK punting on all of that for now to just get basic logging working in Puyolib.

One other request: could you take a look into making the necessary changes for ClientNG? If not, I can take a look. I feel we really ought not merge something that breaks the overall build, and I think this shouldn't be too hard to get working.

File: Client/gamemanager.cpp (hunk 2 above)
Comment (LossRain) [reply Client/gamemanager.cpp]: Addressed, corrected. A classic case of IDE automatically micro-optimizing.

File: Puyolib/Game.cpp (hunk 5 above)
Comment (LossRain) [reply Puyolib/Game.cpp]: Corrected, removed. It was originally there to check for double-free and I forgot to remove it.

File: Puyolib/DebugLog.h (hunk 7 above)
Comment (LossRain) [reply Puyolib/DebugLog.h]: Done and done.

**Comment by LossRain** [comment]
All conversations were addressed and I am thus re-requesting a review. Feature should now work in ClientNG as well as the original Client. Converted all current debug calls in Puyolib to either DebugLog or to Game::debugText and conditioned them under the already existing Puyolib::debugMode. 

The biggest block currently is getting the Windows action to compile the module as is since it seems to have something against it syntax-wise with no clear errors leading to anything violating the spec.

File: Puyolib/DebugLog.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+#include <functional>
+
+
+namespace ppvs {
+
+using std::function;
+
+enum class DebugMessageType {
+	NONE = 0,
+	ERROR = 1,
```
Comment (jchv) [inline Puyolib/DebugLog.h:18]: My suspicion for why there is an error here on Windows is because something (probably a transitive dependency of `Windows.h`) is defining `ERROR` before including this header. Unfortunately, if it is `Windows.h`, the path of least pain is to just avoid keywords that are defined as Windows preprocessor macros.

**Review by jchv — APPROVED** [review]
OK, thanks. Now that Windows is passing, i think this LGTM.
